tools/linter/adapters/test_device_bias_linter.py

```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python3
+"""
+This lint verifies that every Python test file (file that matches test_*.py or
+*_test.py in the test folder) has a cuda hard code in `requires_gpu()`
+decorated function to ensure that the test not fail on other GPU.
+
+"""
+
+from __future__ import annotations
+
+import argparse
+import ast
+import json
+import multiprocessing as mp
+from enum import Enum
+from typing import NamedTuple
+
+
+LINTER_CODE = "TEST_DEVICE_BIAS"
+
+
+class LintSeverity(str, Enum):
+    ERROR = "error"
+    WARNING = "warning"
+    ADVICE = "advice"
+    DISABLED = "disabled"
+
+
+class LintMessage(NamedTuple):
+    path: str | None
+    line: int | None
+    char: int | None
+    code: str
+    severity: LintSeverity
+    name: str
+    original: str | None
+    replacement: str | None
+    description: str | None
+
+
+DEVICE_BIAS = "cuda"
+
+
+class DeviceBiasVisitor(ast.NodeVisitor):
+    def __init__(self, filename: str):
+        self.filename = filename
+        self.lint_messages: list[LintMessage] = []
+
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
+        # Check if the function is decorated with @requires_gpu, which indicates
+        # that the function is intended to run on GPU devices (e.g., CUDA or XPU),
+        # but ensure it does not hardcode the device to CUDA.
+        has_requires_gpu = any(
+            (isinstance(d, ast.Name) and d.id == "requires_gpu")
+            or (
+                isinstance(d, ast.Call)
+                and (isinstance(d.func, ast.Name) and d.func.id == "requires_gpu")
+            )
+            for d in node.decorator_list
+        )
+        if has_requires_gpu:
+            msg_prefix = "`@requires_gpu` function should not hardcode"
+            for subnode in ast.walk(node):
+                if isinstance(subnode, ast.keyword):
+                    if subnode.arg == "device":
+                        val = subnode.value
+                        if isinstance(val, ast.Constant) and DEVICE_BIAS in val.value:
+                            # detect: device="cuda"
+                            self.record(subnode, f"{msg_prefix} device='{val.value}'")
+                        elif isinstance(val, ast.Call):
+                            # detect: device=torch.device("cuda")
+                            if (
+                                isinstance(val.func, ast.Attribute)
+                                and val.func.attr == "device"
+                                and len(val.args) > 0
+                                and isinstance(val.args[0], ast.Constant)
+                                and DEVICE_BIAS in val.args[0].value
+                            ):
+                                self.record(
```

Comment:
The code block is deeply nested. It would be helpful to simplify or refactor it to improve clarity.

Reply:
> The code block is deeply nested. It would be helpful to simplify or refactor it to improve clarity.

Refined, please have a review.